sql/stats: enable a couple of tests under test tenants

yuzefovich opened the pull request:
**sql/stats: prevent deadlocks in a couple of tests when failing**

The deferred functions that close the channel (which can block the
server) need to be executed _before_ `Stopper.Stop` of that server.

**sql/stats: enable a couple of tests under test tenants**

`TestAtMostOneRunningCreateStats` was subtly broken. Namely, we ignored
the error at the end of the test on the "background" full stats job.
This was added in 7de0ae1c607c1665e98d88b61a4008d9940806ea when we
introduced the logic to treat the paused auto stats job as blocking any
other concurrent auto stats jobs. This commit improves the test by
pausing the job in 50% cases (in other 50% we simply block the
background stats job, similar to how things were when the test was first
added in 048bd3d6a0e18217f1b95e579b604f332f3e024b) and then adjusting the
expectations around the error sent on the channel. In particular, if the
job is paused, then we expect to see an error even though the table
stats are collected.

`TestStatsAreDeletedForDroppedTables` and `TestBackgroundAutoPartialStats` simply worked.

Fixes: #109379.
Fixes: #109380.
Epic: CRDB-48945
Release note: None

yuzefovich committed:
sql/stats: prevent deadlocks in a couple of tests when failing

The deferred functions that close the channel (which can block the
server) need to be executed _before_ `Stopper.Stop` of that server.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/155825)
<!-- Reviewable:end -->

yuzefovich committed:
sql/stats: enable a couple of tests under test tenants

`TestAtMostOneRunningCreateStats` was subtly broken. Namely, we ignored
the error at the end of the test on the "background" full stats job.
This was added in 7de0ae1c607c1665e98d88b61a4008d9940806ea when we
introduced the logic to treat the paused auto stats job as blocking any
other concurrent auto stats jobs. This commit improves the test by
pausing the job in 50% cases (in other 50% we simply block the
background stats job, similar to how things were when the test was first
added in 048bd3d6a0e18217f1b95e579b604f332f3e024b) and then adjusting the
expectations around the error sent on the channel. In particular, if the
job is paused, then we expect to see an error even though the table
stats are collected.

`TestStatsAreDeletedForDroppedTables` and
`TestBackgroundAutoPartialStats` simply worked.

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @mgartner reviewed 1 of 1 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155825#-:-OcgclqbEIEDXf3e7b5-:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)



<!-- Sent from Reviewable.io -->
